### PR TITLE
Add ManagedAgent changes to SubmitTaskStateChange

### PR DESCRIPTION
**Pull request by angelcar**
### Summary

This PR covers the changes required to send managed agent state changes to FE

### Implementation details

* Added managed agent state change models where appropriate 
* Sending managed agent state changes follows the same pattern used to send container state changes
  * Changes are batched as container state changes inside of a task state change (task state change is the only supported way of sending any state changes)
### Testing

* Did some manual testing to verify that changes are being sent to FE
* Added UTs to cover new code
* Modified/expanded existing UTs where appropriate 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/eventhandler/task_handler_types.go [hunk 1]
```diff
@@ -235,10 +236,23 @@ func updataTaskSentStatus(task *apitask.Task, status apitaskstatus.TaskStatus, d
...
+		ma, ok := container.GetManagedAgentByName(mac.Name)
+		if ok && ma.SentStatus < mac.Status {
+			saveContainer = true
+			container.UpdateManagedAgentSentStatus(ma.Name, mac.Status)
+		}
+	}
+	if saveContainer {
+		err := dataClient.SaveContainer(container)
+		if err != nil {
+			seelog.Errorf("Failed to update container sent status in databse for container %s: %v", container.Name, err)
```
Comment (fierlion) [inline agent/eventhandler/task_handler_types.go:255]: nit `status in database`

File: agent/statechange/statechange.go [hunk 2]
```diff
@@ -25,6 +25,10 @@ const (
 	// AttachmentEvent is used to define the attachment state transition events
 	// emitted by ENI watcher
 	AttachmentEvent
+
+	// ManagedAgentEvent is used to define the managed agent state transition events
+	// emitted by the engine
+	ManagedAgentEvent
```
Comment (fierlion) [inline agent/statechange/statechange.go:31]: glad to see this differentiation between `ContainerEvent` and `ManagedAgentEvent`

**Review by fierlion — APPROVED** [review]
I like it.

File: agent/eventhandler/task_handler_types.go (hunk 1 above)
Comment (angelcar) [reply agent/eventhandler/task_handler_types.go]: eagle eye! changed.

**Review by ubhattacharjya — APPROVED** [review]

File: agent/eventhandler/task_handler.go [hunk 3]
```diff
@@ -221,6 +223,31 @@ func (handler *TaskHandler) taskStateChangesToSend() []api.TaskStateChange {
...
 func (handler *TaskHandler) batchContainerEventUnsafe(event api.ContainerStateChange) {
+	// if this is a managed agent state change, let's try to bundle it in an existing container event
+	if event.GetEventType() == statechange.ManagedAgentEvent {
+		containerChanges, ok := handler.tasksToContainerStates[event.TaskArn]
+		if ok {
+			for i, cs := range containerChanges {
+				if cs.ContainerName == event.ContainerName && cs.Status == event.Status {
+					cs.ManagedAgents = append(cs.ManagedAgents, event.ManagedAgents...)
+					containerChanges[i] = cs
+					return
```
Comment (sharanyad) [inline agent/eventhandler/task_handler.go:234]: just to understand, we will not send a container event with just a managed agent state change if there is an existing container event in queue?

File: agent/eventhandler/task_handler_types.go [hunk 4]
```diff
@@ -235,10 +236,23 @@ func updataTaskSentStatus(task *apitask.Task, status apitaskstatus.TaskStatus, d
...
-		seelog.Errorf("Failed to update container sent status in databse for container %s: %v", container.Name, err)
+func updateContainerSentStatus(container *apicontainer.Container, status apicontainerstatus.ContainerStatus, managedAgentChanges []api.ManagedAgentStateChange, dataClient data.Client) {
+	saveContainer := false
+	if container.GetSentStatus() < status {
+		saveContainer = true
+		container.SetSentStatus(status)
+	}
+	for _, mac := range managedAgentChanges {
+		ma, ok := container.GetManagedAgentByName(mac.Name)
+		if ok && ma.SentStatus < mac.Status {
```
[the remark this answers is not in the document]
Comment (angelcar) [reply agent/eventhandler/task_handler_types.go:247]: In this case the whole `ManagedAgent` retrieved by `GetManagedAgentByName` is a read only copy (therefore concurrency safe), I don't think a lock is needed. 
What is the risk that you see here?

File: agent/eventhandler/task_handler.go (hunk 3 above)
Comment (angelcar) [reply agent/eventhandler/task_handler.go:234]: That's correct. You can think of it as one more level of batching: managed agent changes are batched in container changes, and container changes is batched in task changes.

**Pull request merged**